### PR TITLE
Reorganize the Slots API for recipes.

### PR DESCRIPTION
**Pull request by shans**
I'm trying to make it more consistent with handles and slandles, and also wanted to clean up the logic in match-recipe-by-verb a bit.

### CODE REVIEW

File: src/runtime/arcs-types/particle-spec.ts [hunk 1]
```diff
@@ -381,6 +381,18 @@ export class ParticleSpec {
...
+  /**
+   * Returns true if there's a consume or provide connection named @name.
+   */
+  hasSlotConnectionName(name: string): boolean {
+    if (this.slotConnections.has(name)) {
+      return true;
+    }
+    return [...this.slotConnections.values()].map(connection => {
+      return connection.provideSlotConnections.map(pc => pc.name).includes(name);
+    }).reduce((a, b) => a || b);
```
Comment (Cypher1) [inline src/runtime/arcs-types/particle-spec.ts:393]: @shans will the node+chrome optimisers actually be able to convert this to an early exit? I'm not seriously considering the performance of this, but it does seem like the kind of situation where the underlying impl should just be a look up on a map from name to slot.

File: src/runtime/recipe/internal/recipe-walker.ts [hunk 2]
```diff
@@ -66,8 +66,7 @@ export class RecipeWalker extends Walker<Recipe> {
     }
     if (this.onPotentialSlotConnection) {
       for (const particle of recipe.particles) {
-        for (const [name, slotSpec] of particle.getSlotSpecs()) {
-          if (particle.getSlotConnectionByName(name)) continue;
+        for (const slotSpec of particle.getUnboundSlots()) {
```
Comment (Cypher1) [inline src/runtime/recipe/internal/recipe-walker.ts:69]: Will this restrict a particle from rendering to two different slots?

Also does this suggest that `particle.getUnboundSlots` should actually be `particle.getUnboundSlotConnections`? (wordy, but more accurate)

**Review by Cypher1 — APPROVED** [review]
LGTM (with a question / suggestion that doesn't need to be addressed here)

File: src/runtime/recipe/internal/recipe-walker.ts (hunk 2 above)
Comment (shans) [reply src/runtime/recipe/internal/recipe-walker.ts]: This should be an isomorphic change; just moving the filter from here into particle to make the API for slot connections look more like the API for handle connections. I think you're right about the naming.

**Pull request merged**